apply ACPERM rules in order (#427)
riscv/riscv-cheri: src/level-ext.adoc

11+| bit[0] - <<m_bit>> ({CAP_MODE_VALUE}-{cheri_cap_mode_name}, {INT_MODE_VALUE}-{cheri_int_mode_name})
|Bits[4:3]| R | W | C | LM | EL | SL  | X | ASR | Mode^1^ |
| 0-1   | ✔ | ✔ | ✔ | ✔  | ✔  | ∞   | ✔ |  ✔  | Mode^1^  | Execute + ASR (see <<infinite-cap>>)
| 2-3   | ✔ |   | ✔ | ✔  | ✔  | ∞^1^| ✔ |     | Mode^1^  | Execute + Data & Cap RO
| 4-5   | ✔ | ✔ | ✔ | ✔  | ✔  | ∞   | ✔ |     | Mode^1^  | Execute + Data & Cap RW

> | 6-7   | ✔ | ✔ |   |    |    | ∞^1^| ✔ |     | Mode^1^  | Execute + Data RW

Reviewer: This could be 0 or ∞, because it truly doesn't matter as we will never load a writable cap via this (bearing future transitive SL in mind). 0 might be slightly friendlier. One could imagine a structure made deeply not SL, with some function pointers in it. Those should be sealed, which is why I don't really think this will ever be bothersome. But if there were any unsealed ones, and they had SL reduced, I think it would be more expected that they lose C, not X, as a result (possibly, they really should be sealed, so who knows what is more useful there).

Given that it is probably not important, and the rule "lose X if SL not inf" is easier than something like: "lose X if SL if inf != W", I don't have strong feelings here. Maybe somebody else has thoughts here?

Author: I have changed it. And we've also changed the order of rules 9 and 10, and update (former) rule 9 to allow for SL=0 in this case.

Previously it was testing for SL (former rule 9) and then clearning SL (former rule 10) which seems the wrong way around

With the latest commit if you start with infinite permissions, just remove C then you stay in quadrant 1 and end on 6-7, and so retain the X permission.

Previously the X was lost which seemed a shame, as it could have been kept.

11+| *Quadrant 2: Restricted capability data read/write*
11+| bit[2] = write, bit[1:0] = store level. R and C implicitly granted, LM dependent on W permission.
|Bits[4:3]| R | W | C | LM | EL | SL    | X | ASR | Mode^1^ |
| 0-2   10+| reserved

> | 3       | ✔ |   | ✔ |    |    | ∞^1^  |   |     | N/A | Data & Cap R0 (without <<lm_perm>>)

Reviewer 2: ```suggestion
| 3       | ✔ |   | ✔ |    |    | 0^1^  |   |     | N/A | Data & Cap R0 (without <<lm_perm>>)
```

Reviewer 2: IMO this should be zero for consistency: just report the negation of bit 0 (and bit 0-1 for multiple level bits).

Reviewer: I don't think it hurts to have 0 here (if as Alex says it adds consistency). Given this cap does not have LM, it will never load anything to which SL (even transitively) would ever apply.

11+| _Reserved bits for future extensions must be 1 so they are implicitly granted_
|Bits[4:3]| R | W | C | LM | EL | SL    | X | ASR | Mode^1^ |
| 0-2   10+| reserved

> | 3       | ✔ |   | ✔ | ✔  | ✔  | ∞^1^  |   |     | N/A | Data & Cap R0

Reviewer 2: ```suggestion
| 3       | ✔ |   | ✔ | ✔  | ✔  | 0^1^  |   |     | N/A | Data & Cap R0
```

Reviewer 2: IMO this should be zero for consistency: just report the negation of bit 0 (and bit 0-1 for multiple level bits).

Reviewer: This one can load a W cap, and so its SL would be important if we add a transitive SL. In an earlier sketch we had all combinations of SL and W=0 in this quadrant for this reason.

I think we decided it was fine to reserve 3/4 of these now (and just have the one, 3 or 0). Whichever one we have selected here with the bit pattern is what we should report. I think that is (as Alex suggests) 0, to match the W=1 case. We can add 1/2/3 if required. This is slightly unlike Quad 1, where I expect to have infinity regardless of how many levels there are.

Reviewer 2: Yes I agree if we end up adding the transitive property, then 2 should report SL=1, and for n bit we add 1 with SL=2 and zero with SL=3.

| 4       | ✔ | ✔ | ✔ | ✔  | ✔  | _(3)_ |   |     | N/A | Reserved for `LVLBITS=2`
| 5       | ✔ | ✔ | ✔ | ✔  | ✔  | _(2)_ |   |     | N/A | Reserved for `LVLBITS=2`
| 6       | ✔ | ✔ | ✔ | ✔  | ✔  | 1     |   |     | N/A | Data & Cap RW (with store _local_)
| 7       | ✔ | ✔ | ✔ | ✔  | ✔  | 0     |   |     | N/A | Data & Cap RW (no store _local_)
|==============================================================================

> ^1^ SL isn't applicable in these cases, but is specified as ∞ to help the definition of <<ACPERM>>

Reviewer 2: ```suggestion
^1^ SL isn't applicable in these cases, but reports this value in <<GCPERM>> to simplify the rules followed by <<ACPERM>>
```

Maybe something like this so we can report 0 for the Q2&Q3 cases.
